_posts/2023-04-19-more-optimizations.md

```diff
@@ -0,0 +1,1421 @@
+
+With the guard test, the first BEAM instruction in `expand/2` is:
+
+```
+    {bs_start_match4,{atom,no_fail},2,{x,0},{x,0}}.
+```
+
+Without the guard test, the first BEAM instruction is:
+
+```
+    {test,bs_start_match3,{f,3},2,[{x,0}],{x,2}}.
+```
+
+The `bs_start_match4` instruction is more efficient because it does
+not have to test that `{x,0}` contains a binary.
+
+The benchmark results show measurable increased execution time for
+`expand/1` if the guard test is removed:
+
+```
+erlperf --init_runner_all 'rand:bytes(10_000).' \
+        'r(Bin) -> append:expand(Bin).' \
+        'r(Bin) -> append:expand_bc(Bin).'
+Code                                     ||        QPS       Time   Rel
+r(Bin) -> append:expand_bc(Bin).          1      13273   75366 ns  100%
+r(Bin) -> append:expand(Bin).             1      11875   84236 ns   89%
+```
+
+### Revisiting the `base64` module
+
+Traditionally, up to OTP 25, the clause in the `base64` module that does
+most of the work of encoding a binary to Base64 looked like this:
+
+```erlang
+encode_binary(<<B1:8, B2:8, B3:8, Ls/bits>>, A) ->
+    BB = (B1 bsl 16) bor (B2 bsl 8) bor B3,
+    encode_binary(Ls,
+                  <<A/bits,(b64e(BB bsr 18)):8,
+                    (b64e((BB bsr 12) band 63)):8,
+                    (b64e((BB bsr 6) band 63)):8,
+                    (b64e(BB band 63)):8>>).
+```
+
+The reason is that matching out segments of size 8 has always been
+specially optimized and has been much faster than matching out a
+segment of size 6. That is no longer true in OTP 26. With the
+improvements in binary matching described in this blog post, the
+clause can be written in a more natural way:
+
+```erlang
+encode_binary(<<B1:6, B2:6, B3:6, B4:6, Ls/bits>>, A) ->
+    encode_binary(Ls,
+                  <<A/bits,
+                    (b64e(B1)):8,
+                    (b64e(B2)):8,
+                    (b64e(B3)):8,
+                    (b64e(B4)):8>>);
+```
+
+(This is not the exact code in OTP 26, because of
+[additional](https://github.com/erlang/otp/pull/6280)
+[features](https://github.com/erlang/otp/pull/6711) added later.)
+
+The benchmark results for encoding a random binary of 1,000,000 bytes
+to Base64 for OTP 25 is:
+
+```
+erlperf --init_runner_all 'rand:bytes(1_000_000).' \
+        'r(Bin) -> base64:encode(Bin).'
+Code                                  ||        QPS       Time
+r(Bin) -> base64:encode(Bin).          1         61   16489 us
+```
+
+The benchmark results for encoding a random binary of 1,000,000 bytes
+to Base64 for OTP 26 is:
+
+```
+erlperf --init_runner_all 'rand:bytes(1_000_000).' \
+        'r(Bin) -> base64:encode(Bin).'
+Code                                  ||        QPS       Time
+r(Bin) -> base64:encode(Bin).          1        249    4023 us
+```
+
+That is, encoding is about 4 times faster.
+
+### Please try this at home!
+
+If you want to follow along and examine the native code for loaded
+modules, start the runtime system like this:
```

Comment:
At the start of the post I thought that it would be nice to be told how to dump the generated assembly, so perhaps this should be moved much earlier?

Reply:
You are the second one suggesting it (https://github.com/bjorng/erlang-org/pull/2#discussion_r1155685456), so I will move the section and add some introduction to avoid scaring away people.